agent/taskresource/asmsecret/asmsecret.go

```diff
@@ -0,0 +1,426 @@
+}
+
+// SetCreatedAt sets the timestamp for resource's creation time
+func (secret *ASMSecretResource) SetCreatedAt(createdAt time.Time) {
+	if createdAt.IsZero() {
+		return
+	}
+	secret.lock.Lock()
+	defer secret.lock.Unlock()
+
+	secret.createdAt = createdAt
+}
+
+// GetCreatedAt sets the timestamp for resource's creation time
+func (secret *ASMSecretResource) GetCreatedAt() time.Time {
+	secret.lock.RLock()
+	defer secret.lock.RUnlock()
+
+	return secret.createdAt
+}
+
+// It spins up multiple goroutines in order to retrieve values in parallel.
+func (secret *ASMSecretResource) Create() error {
+
+	// To fail fast, check execution role first
+	executionCredentials, ok := secret.credentialsManager.GetTaskCredentials(secret.getExecutionCredentialsID())
+	if !ok {
+		// No need to log here. managedTask.applyResourceState already does that
+		err := errors.New("asm secret resource: unable to find execution role credentials")
+		secret.setTerminalReason(err.Error())
+		return err
+	}
+	iamCredentials := executionCredentials.GetIAMRoleCredentials()
+
+	var wg sync.WaitGroup
+
+	// Get the maximum number of errors be returned, which will be one error per goroutine
+	errorEvents := make(chan error, len(secret.requiredSecrets))
+
+	seelog.Infof("asm secret resource: retrieving secrets for containers in task: [%s]", secret.taskARN)
+	secret.secretData = make(map[string]string)
+
+	for _, asmsecret := range secret.getRequiredSecrets() {
+		wg.Add(1)
+		// Spin up goroutine per secret to speed up processing time
+		go secret.retrieveASMSecretValue(asmsecret, iamCredentials, &wg, errorEvents)
+	}
+
+	wg.Wait()
+	close(errorEvents)
+
+	if len(errorEvents) > 0 {
+		var terminalReasons []string
+		for err := range errorEvents {
+			terminalReasons = append(terminalReasons, err.Error())
+		}
+
+		errorString := strings.Join(terminalReasons, ";")
+		secret.setTerminalReason(errorString)
+		return errors.New(errorString)
+	}
+
+	return nil
+}
+
+// retrieveASMSecretValue reads secret value from cache first, if not exists, call GetSecretFromASM to retrieve value
+// AWS secrets Manager
+func (secret *ASMSecretResource) retrieveASMSecretValue(apiSecret apicontainer.Secret, iamCredentials credentials.IAMRoleCredentials, wg *sync.WaitGroup, errorEvents chan error) {
+	defer wg.Done()
+
+	asmClient := secret.asmClientCreator.NewASMClient(apiSecret.Region, iamCredentials)
+	seelog.Infof("asm secret resource: retrieving resource for secret %v in region %s for task: [%s]", apiSecret.ValueFrom, apiSecret.Region, secret.taskARN)
+	secretValue, err := asm.GetSecretFromASM(apiSecret.ValueFrom, asmClient)
+	if err != nil {
+		errorEvents <- fmt.Errorf("fetching secret data from AWS Secrets Manager in region %s: %v", apiSecret.Region, err)
+		return
+	}
+
+	secret.lock.Lock()
+	defer secret.lock.Unlock()
+
+	// put secret value in secretData
+	secretKey := apiSecret.GetSSMSecretResourceCacheKey()
+	secret.secretData[secretKey] = secretValue
```

Comment:
Thoughts on using a sync map instead of locking/unlocking?

https://golang.org/doc/go1.9#sync-map